File: python/binarySearchTree/binarySearchTree.py
from node import Node

class BinarySearchTree:
    root = None

    def toString(self, root, arr):
        if root is None:
            return

        self.toString(root.left, arr)
        arr.append(root.key)
        self.toString(root.right, arr)

    def find(self, key, node):
        if key == node.key:
            return node

        if key > node.key:
            if node.right is None:
                return node
            return self.find(key, node.right)
        else:
            if node.left is None:
                return node
            return self.find(key, node.left)

    def next(self, node):
        if node.right is not None:
            return self._leftDescendant(node.right)
        else:
            return self._rightAncestor(node.key, node.parent)

    def _leftDescendant(self, node):
        if node.left is not None:
            return self._leftDescendant(node.left)
        return node

    def _rightAncestor(self, key, node):
        if node is None:
            return None
        
        if key < node.key:
            return node
        else:
            return self._rightAncestor(key, node.parent)

    def rangeSearch(self, st, ed, root):
        result = []
        node = self.find(st, root)

        while (node is not None and node.key <= ed):
            if node.key >= st:
                result.append(node.key)
            node = self.next(node)

        return result

    def insert(self, key, root):
        if root is None:
            return Node(key, None)

        node = self.find(key, root)
        newNode = Node(key, node)

        if key < node.key:
            node.left = newNode
        else:
            node.right = newNode

        return root

    def remove(self, node):
        if node.right is None:
            l = node.left
            p = node.parent

            if l is None:
                self.root = None
                return
            
            l.parent = p

            if p is not None:
                if p.left is not None and p.left.key == node.key:
                    p.left = l
                if p.right is not None and p.right.key == node.key:
                    p.right = l
            else:
                self.root = l

        else:
            nextNode = self.next(node)
            node.key = nextNode.key

            p = nextNode.parent
            r = nextNode.right

            if p is not None:
                if p.left is not None and p.left.key == nextNode.key:
                    p.left = r
                if p.right is not None and p.right.key == nextNode.key:
                    p.right = r
            
            if r is not None:
                r.parent = p

File: java/src/binarySearchTree/BinarySearchTree.java
package binarySearchTree;

import java.util.ArrayList;
import java.util.List;

public class BinarySearchTree {
    public Node root;

    public List<Integer> toString(Node node, List<Integer> result) {
        if (node == null)
            return result;

        this.toString(node.left, result);
        result.add(node.key);
        this.toString(node.right, result);

        return result;
    }

    public Node find(int key, Node node) {
        if (node.key == key)
            return node;

        if (node.key > key) {
            if (node.left == null)
                return node;
            return this.find(key, node.left);
        } else {
            if (node.right == null)
                return node;
            return this.find(key, node.right);
        }
    }

    public Node next(Node node) {
        if (node.right != null)
            return this.leftDescendant(node.right);
        else
            return this.rightAncestor(node.key, node.parent);
    }

    private Node leftDescendant(Node node) {
        if (node.left != null)
            return this.leftDescendant(node.left);
        return node;
    }

    private Node rightAncestor(int key, Node node) {
        if (node == null)
            return null;

        if (key < node.key)
            return node;
        else
            return this.rightAncestor(key, node.parent);
    }

    public Integer[] rangeSearch(int st, int ed, Node root) {
        List<Integer> result = new ArrayList<Integer>();

        Node node = this.find(st, root);
        while(node != null && node.key <= ed) {
            if (node.key >= st)
                result.add(node.key);
            node = this.next(node);
        }

        return result.toArray(new Integer[result.size()]);
    }

    public Node insert(int key, Node root) {
        if (root == null)
            return new Node(key, null);

        Node node = this.find(key, root);
        Node newNode = new Node(key, node);

        if (key < node.key)
            node.left = newNode;
        else
            node.right = newNode;

        return root;
    }

    public void remove(Node node) {
        if (node.right == null) {
            // if single node tree
            if (node.left == null) {
                this.root = null;
                return;
            }

            // level up node.left and remove node
            node.left.parent = node.parent;

            if (node.parent != null) {
                if (node.parent.left.key == node.key) {
                    node.parent.left = node.left;
                } else {
                    node.parent.right = node.left;
                }
            } else {
                this.root = node.left;
            }
        } else {
            // find left most node
            Node next = this.next(node);

            // assign key: node <- next
            node.key = next.key;

            // level up next.right and remove next
            if (next.right != null)
                next.right.parent = next.parent;

            if (next.parent.left != null && next.parent.left.key == next.key)
                next.parent.left = next.right;

            if (next.parent.right != null && next.parent.right.key == next.key)
                next.parent.right = next.right;
        }
    }
}
